examples/quantum_variation.py

```diff
@@ -0,0 +1,321 @@
+import math
+import numpy as np
+import matplotlib.pyplot as plt
+import tensorcircuit as tc
+
+tc.set_backend("tensorflow")
+tc.set_dtype("complex64")
+
+# realize R gates in paper
+def R_gate(k, c, ODE_theta):
+    if door[k][0] == 0:
+        c.rx(door[k][1] + 1, theta=ODE_theta[k])
+    if door[k][0] == 1:
+        c.ry(door[k][1] + 1, theta=ODE_theta[k])
+    if door[k][0] == 2:
+        c.rz(door[k][1] + 1, theta=ODE_theta[k])
+    if door[k][0] == 3:
+        c.rxx(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 4:
+        c.ryy(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 5:
+        c.rzz(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 6:
+        c.crx(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 7:
+        c.cry(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+    if door[k][0] == 8:
+        c.crz(door[k][1] + 1, door[k][2] + 1, theta=ODE_theta[k])
+
+
+# realize U gates in paper
+def U_gate_conditional(k):
+    if door[k][0] == 0:
+        gate_now = tc.gates.multicontrol_gate(
+            np.kron(tc.gates._x_matrix, np.eye(2)), [1]
+        )
+    if door[k][0] == 1:
+        gate_now = tc.gates.multicontrol_gate(
+            np.kron(tc.gates._y_matrix, np.eye(2)), [1]
+        )
+    if door[k][0] == 2:
+        gate_now = tc.gates.multicontrol_gate(
+            np.kron(tc.gates._z_matrix, np.eye(2)), [1]
+        )
+    if door[k][0] == 3:
+        gate_now = tc.gates.multicontrol_gate(tc.gates._xx_matrix, [1])
+    if door[k][0] == 4:
+        gate_now = tc.gates.multicontrol_gate(tc.gates._yy_matrix, [1])
+    if door[k][0] == 5:
+        gate_now = tc.gates.multicontrol_gate(tc.gates._zz_matrix, [1])
+    return gate_now.eval_matrix()
+
+
+# realize Hamilton gates in ancillary circuit
+def H_gate_conditional(q):
+    if h_door[q][0] == 0:
+        gate_now = tc.gates.multicontrol_gate(
+            np.kron(tc.gates._x_matrix, np.eye(2)), [1]
+        )
+    if h_door[q][0] == 1:
+        gate_now = tc.gates.multicontrol_gate(
+            np.kron(tc.gates._y_matrix, np.eye(2)), [1]
+        )
+    if h_door[q][0] == 2:
+        gate_now = tc.gates.multicontrol_gate(
+            np.kron(tc.gates._z_matrix, np.eye(2)), [1]
+        )
+    if h_door[q][0] == 3:
+        gate_now = tc.gates.multicontrol_gate(tc.gates._xx_matrix, [1])
+    if h_door[q][0] == 4:
+        gate_now = tc.gates.multicontrol_gate(tc.gates._yy_matrix, [1])
+    if h_door[q][0] == 5:
+        gate_now = tc.gates.multicontrol_gate(tc.gates._zz_matrix, [1])
+    return gate_now.eval_matrix()
```

Comment:
there is no diff in U gate and H gate method? merge to one method?

Reply:
but the parameter change from door to h_door

Reply:
in this case, why dont you create a method with two arguments: ``U_gate(door, q)``

Reply:
perhaps it works slowly if I set new parameter in wrapping my fun with vmap

Reply:
I dont think so, you can have a comparison